### PR TITLE
U20 LastMatch module (Eligibility for U20 players)

### PR DESCRIPTION
**Pull request by rferromoreno**
I'm proudly doing my first contribution to Foxtrick. In this time, I'm adding a module for checking which is the last official match that any player under 20 years can play.

A graphical example can be found on request #1542 where I posted two screenshots.

Pages that this module involves:

- SeniorPlayerDetails
- YouthPlayerDetails
- AllPlayersDetails

Related issues: 
#893
#1542

### CODE REVIEW

**Comment by minj** [comment]
This change is [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)

**Comment by rferromoreno** [comment]
There is a `RomanToDec` method on `util/id.js`, but not a `DecToRoman`. If you want to, I can put this method on the `id.js` module, but I think it fit better on `misc.js`.

I just did the string replace pattern thing and I'm pushing that change :)

**Comment by minj** [comment]
Sorry about the delay. I shouldn't be surprised any more, really, but I still somehow manage to overestimate how much juice I have left on weekdays.

**Comment by minj** [comment]
It would be nice to move the roman numeral methods to misc.js. And a better implementation of `romanToDec` is welcome, as well. Provided we can actually use it.

Ideas for improvement: add this to transfers and the relevant SkillTable views.

It's working fine, obviously, but I have some more corrections/nitpicks.
Detailed review comments will follow:

---

Reviewed 8 of 10 files at r1.
Review status: 8 of 10 files reviewed at latest revision, all discussions resolved.

---

*[content/foxtrick.properties, line 518 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBAkQM694ysFRgdqsvQ:-LBAkQM694ysFRgdqsvR:b-u6w1m9) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/foxtrick.properties#L518)):*
> ```INI
> 
> # U20LastMatch
> module.U20LastMatch.desc=Show which is the last U20 official match that an U20 can play.
> ```

Nitpick, here and others: Show **the** last official U20 match that a **player could** play in.

---

*[content/foxtrick.properties, line 1763 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBAl1I49hp0DfaCE6pv:-LBAl1I49hp0DfaCE6pw:b-iyvhz2) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/foxtrick.properties#L1763)):*
> ```INI
> # U20LastMatch
> U20LastMatch.title=Last U20 match:
> U20LastMatch.worldcup=WC 
> ```

whitespace

---

*[content/information-aggregation/u20-lastmatch.js, line 14 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBOfTCAC_0Knv5CLF3:-LBBOfTCAC_0Knv5CLF4:b-i8xml5) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L14)):*
> ```JavaScript
> 		'YouthPlayers', 'SeniorPlayers', 'AllPlayers'
> 	],
>   MATCHES_OFFSETS: [
> ```

The way I see it, if the day value is less than `i`th element, you can only play `i`th game (where i starts at 0).

Wouldn't it be better to call this something like `DATE_CUTOFFS`?

At the very least a comment would be nice.

---

*[content/information-aggregation/u20-lastmatch.js, line 79 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBApa32E0BsCkLVAz_x:-LBApa32E0BsCkLVAz_y:b-ex2akk) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L79)):*
> ```JavaScript
>       roman = "",
>       i = 3;
>     while (i--) roman = (key[+digits.pop() + i * 10] || "") + roman;
> ```

this is kinda clever, I guess, is this your own, or you found it on SO or some place else?

---

*[content/information-aggregation/u20-lastmatch.js, line 98 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBB6ZiB-wQjv_j5c0K9:-LBB6ZiB-wQjv_j5c0KA:bhyb1f2) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L98)):*
> ```JavaScript
> 
>     // Round I, Match #1 of World Cup XXVI. (26/05/2017)
>     var origin = Foxtrick.util.time.toHT(doc, new Date(2017, 4, 26));
> ```

Timezones are a mess, even more so in HT, even more so in FT.

The age you get is in HT time and flips at midnight CET IIRC. However, dates displayed are in user time. And JS (browser) time might be different still.

Since we are forced to use string parsing in `get(HT)Date`, and then again serialization in `buildDate`, the timestamps used are actually in browser timezone anyway but do not (generally) equal to `Date.now()`.

This method expects a user date, like returned from `getDate`, not a browser date, which is generated from `Date` constructor with specific params. 
Long story short, all timestamps that do not refer to UTC can be compared directly with the Date constructor, if you use a timestamp for the same 'timezone' as the original timestamp (i.e. HT time):

`var origin = new Date(2017, 4, 26);`

Of course, DST causes issues here and I have old plans to import a proper timezone lib but alas...

---

*[content/information-aggregation/u20-lastmatch.js, line 100 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBFyzH9DW3t3oFgTOM:-LBBFyzH9DW3t3oFgTON:bjip71i) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L100)):*
> ```JavaScript
>     var origin = Foxtrick.util.time.toHT(doc, new Date(2017, 4, 26));
>     var msDiff = dateWhenPlayerIs21.getTime() - origin.getTime();
>     var dayDiff = Math.floor(msDiff / MSECS_IN_DAY) - 1;
> ```

it would be nice to explain the -1 with a comment

---

*[content/information-aggregation/u20-lastmatch.js, line 122 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBHPS5Bygpu90ItzAM:-LBBHPS5Bygpu90ItzAN:bevh4o0) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L122)):*
> ```JavaScript
> 
>     var MATCHES_DESCRIPTIONS = [
>       ROUND1.replace(/%2/,"1"),
> ```

nitpick: wouldn't a few for loops be better to populate this? :)

---

*[content/information-aggregation/u20-lastmatch.js, line 189 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBIM9nDQzMP9CoxdmD:-LBBIM9nDQzMP9CoxdmE:b-9lo4oq) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L189)):*
> ```JavaScript
>       // Display the U20 Last Match information.
>       var table = doc.querySelector(".playerInfo table");
>       var row = Foxtrick.insertFeaturedRow(table, module, table.rows.length);
> ```

I especially appreciated that you found and used the `featured` methods on your own

---

*[content/information-aggregation/u20-lastmatch.js, line 194 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBIAAlEK9X82LHjqAn:-LBBIAAlEK9X82LHjqAo:bgmkglh) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L194)):*
> ```JavaScript
>       title.textContent = Foxtrick.L10n.getString("U20LastMatch.title");
>       var lastMatch = row.insertCell(1);
>       lastMatch.textContent = Foxtrick.L10n.getString("U20LastMatch.worldcup");
> ```

this should use replacement as well

---

*[content/information-aggregation/u20-lastmatch.js, line 208 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBGf61Etxv78E81wmh:-LBBGf61Etxv78E81wmi:bgmkglh) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L208)):*
> ```JavaScript
>         Foxtrick.addClass(row, 'ft-u20-last-match');
>         container.textContent = Foxtrick.L10n.getString("U20LastMatch.title") +
>           ' ' + Foxtrick.L10n.getString("U20LastMatch.worldcup") + 
> ```

this should use replacement as well

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557#-:-LBBScD3BMgRVQJI79tD:bn0r954)*

**Comment by minj** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
*[content/information-aggregation/u20-lastmatch.js, line 79 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBApa32E0BsCkLVAz_x:-LBBWOUx5zEX2wdoCR0A:b497osd) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L79)):*
<details><summary><i>Previously, minj (LA-MJ) wrote…</i></summary><blockquote>

this is kinda clever, I guess, is this your own, or you found it on SO or some place else?
</blockquote></details>

Found it on StackOverflow. Was looking for an efficient way to resolve the DecToRoman. I guess this is the best way, even if the code is hard to read.
Already tested it, and it works like a charm

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
*[content/information-aggregation/u20-lastmatch.js, line 194 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBIAAlEK9X82LHjqAn:-LBCAcoC9jX-XaeBfVxi:b-896fix) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L194)):*
<details><summary><i>Previously, minj (LA-MJ) wrote…</i></summary><blockquote>

this should use replacement as well
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
*[content/information-aggregation/u20-lastmatch.js, line 208 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBGf61Etxv78E81wmh:-LBCAeek8i2eyKRQLns5:b-896fix) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L208)):*
<details><summary><i>Previously, minj (LA-MJ) wrote…</i></summary><blockquote>

this should use replacement as well
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
*[content/information-aggregation/u20-lastmatch.js, line 122 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBHPS5Bygpu90ItzAM:-LBCAjUd7oYTFhbossQG:b-896fix) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L122)):*
<details><summary><i>Previously, minj (LA-MJ) wrote…</i></summary><blockquote>

nitpick: wouldn't a few for loops be better to populate this? :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
*[content/information-aggregation/u20-lastmatch.js, line 14 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBBOfTCAC_0Knv5CLF3:-LBCAnC46_EF9YpwBZ6p:b-896fix) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L14)):*
<details><summary><i>Previously, minj (LA-MJ) wrote…</i></summary><blockquote>

The way I see it, if the day value is less than `i`th element, you can only play `i`th game (where i starts at 0).

Wouldn't it be better to call this something like `DATE_CUTOFFS`?

At the very least a comment would be nice.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
*[content/foxtrick.properties, line 1763 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBAl1I49hp0DfaCE6pv:-LBCApdy0CZnsSI2iWq_:b-896fix) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/foxtrick.properties#L1763)):*
<details><summary><i>Previously, minj (LA-MJ) wrote…</i></summary><blockquote>

whitespace
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
*[content/foxtrick.properties, line 518 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBAkQM694ysFRgdqsvQ:-LBCArc76pPxjey2_mpN:b-896fix) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/foxtrick.properties#L518)):*
<details><summary><i>Previously, minj (LA-MJ) wrote…</i></summary><blockquote>

Nitpick, here and others: Show **the** last official U20 match that a **player could** play in.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
*[content/information-aggregation/u20-lastmatch.js, line 98 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBB6ZiB-wQjv_j5c0K9:-LBCB26m5UH-GFVDA9Bk:begl3yk) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L98)):*
<details><summary><i>Previously, minj (LA-MJ) wrote…</i></summary><blockquote>

Timezones are a mess, even more so in HT, even more so in FT.

The age you get is in HT time and flips at midnight CET IIRC. However, dates displayed are in user time. And JS (browser) time might be different still.

Since we are forced to use string parsing in `get(HT)Date`, and then again serialization in `buildDate`, the timestamps used are actually in browser timezone anyway but do not (generally) equal to `Date.now()`.

This method expects a user date, like returned from `getDate`, not a browser date, which is generated from `Date` constructor with specific params. 
Long story short, all timestamps that do not refer to UTC can be compared directly with the Date constructor, if you use a timestamp for the same 'timezone' as the original timestamp (i.e. HT time):

`var origin = new Date(2017, 4, 26);`

Of course, DST causes issues here and I have old plans to import a proper timezone lib but alas...
</blockquote></details>

I will be testing and checking this. It was coded by me like 8 months ago and I knew there could be some problems with timezones

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
Okay, made the fixes you asked for in the code review.

Of course there are interesting ideas about adding this to other HT pages, but since it's my first contribution, I think I will add those on the next versions.

**Comment by minj** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[content/information-aggregation/u20-lastmatch.js, line 79 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBApa32E0BsCkLVAz_x:-LBCZvL605RtrqakaCLk:b-xmrbuh) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L79)):*
<details><summary><i>Previously, rferromoreno (Ricardo Ferro Moreno) wrote…</i></summary><blockquote>

Found it on StackOverflow. Was looking for an efficient way to resolve the DecToRoman. I guess this is the best way, even if the code is hard to read.
Already tested it, and it works like a charm
</blockquote></details>

AFAIK StackOverflow's policy is we need to add a link to the answer next to the code used.

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by rferromoreno** [comment]
*[content/information-aggregation/u20-lastmatch.js, line 79 at r2](https://reviewable.io/reviews/minj/foxtrick/1557#-LBApa32E0BsCkLVAz_x:-LBCaEjFDeAVteZW7Czy:bq8zotr) ([raw file](https://github.com/minj/foxtrick/blob/db59af7ddf6f77aa8d5e46876a4b696c86768d92/content/information-aggregation/u20-lastmatch.js#L79)):*
<details><summary><i>Previously, minj (LA-MJ) wrote…</i></summary><blockquote>

AFAIK StackOverflow's policy is we need to add a link to the answer next to the code used.
</blockquote></details>

Yeah but it's kind of funny, because the StackOverflow reply source it's from another blog...

Stackoverflow link: https://stackoverflow.com/questions/9083037/convert-a-number-into-a-roman-numeral-in-javascript
Original source: http://blog.stevenlevithan.com/archives/javascript-roman-numeral-converter

---


*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by minj** [comment]
Could you add a commit linking to the blog in misc.js?

I'll merge once that is done and make a new beta build

2018-04-28 20:52 GMT+02:00 Ricardo Ferro Moreno <notifications@github.com>:

**Comment by rferromoreno** [comment]
Done :) :shipit:

**Comment by minj** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/minj/foxtrick/1557)*

**Comment by minj** [comment]
found a small bug right away, heh

`allPlayers` page is running everywhere, including NT etc, where it fails now due to different DOM structure. You need to use `players` page instead. Easy mistake to make.

I'll push a quick-fix.
